model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheSession.java

```diff
@@ -1016,40 +1017,70 @@ public Stream<GroupModel> getGroupsByRoleStream(RealmModel realm, RoleModel role
 
     @Override
     public Stream<GroupModel> getTopLevelGroupsStream(RealmModel realm, String search, Boolean exact, Integer first, Integer max) {
-        String cacheKey = getTopGroupsQueryCacheKey(realm.getId() + search + first + max);
-        boolean queryDB = invalidations.contains(cacheKey) || listInvalidations.contains(cacheKey)
+        // TODO the way the cache is implemented, we can't invalidate keys based on only one property matching
```

Comment:
I've seen https://github.com/keycloak/keycloak/pull/27096 and I'm wondering if the new functionality could be applied here. With this change, the first request for `{realm-name}/groups` will populate the cache with all top-level groups even if we asked just for 20 groups. I tested with 10k groups and I didn't see any noticeable change in performance.

Reply:
Spot on! I originally started looking at the other issue to preserve the functionality for pagination here and then ended up realizing it was a bit more than just a feature enhancement